app/post_acknowledgements.go

```diff
@@ -0,0 +1,141 @@
+		switch {
+		case errors.As(nErr, &appErr):
+			return nil, appErr
+		default:
+			return nil, model.NewAppError("SaveAcknowledgementForPost", "app.acknowledgement.save.save.app_error", nil, "", http.StatusInternalServerError).Wrap(nErr)
+		}
+	}
+
+	a.Srv().Go(func() {
+		a.sendAcknowledgementEvent(model.WebsocketEventAcknowledgementAdded, acknowledgement, post)
+	})
+
+	return acknowledgement, nil
+}
+
+func (a *App) DeleteAcknowledgementForPost(c *request.Context, userID, postID string) *model.AppError {
+	post, err := a.GetSinglePost(postID, false)
+	if err != nil {
+		return err
+	}
+
+	channel, err := a.GetChannel(c, post.ChannelId)
+	if err != nil {
+		return err
+	}
+
+	if channel.DeleteAt > 0 {
+		return model.NewAppError("DeleteAcknowledgementForPost", "api.acknowledgement.delete.archived_channel.app_error", nil, "", http.StatusForbidden)
+	}
+
+	oldAck, nErr := a.Srv().Store().PostAcknowledgement().Get(userID, postID)
+
+	if nErr != nil {
+		var nfErr *store.ErrNotFound
+		switch {
+		case errors.As(nErr, &nfErr):
+			return model.NewAppError("GetPostAcknowledgement", "app.acknowledgement.get.app_error", nil, "", http.StatusNotFound).Wrap(nErr)
+		default:
+			return model.NewAppError("GetPostAcknowledgement", "app.acknowledgement.get.app_error", nil, "", http.StatusInternalServerError).Wrap(nErr)
+		}
+	}
+
+	if model.GetMillis()-oldAck.AcknowledgedAt > 5*60*1000 {
+		return model.NewAppError("DeleteAcknowledgementForPost", "api.acknowledgement.delete.deadline.app_error", nil, "", http.StatusForbidden)
+	}
+
+	nErr = a.Srv().Store().PostAcknowledgement().Delete(oldAck)
+
+	if nErr != nil {
+		return model.NewAppError("DeleteAcknowledgementForPost", "app.acknowledgement.delete_all_with_emoji_name.get_acknowledgement.app_error", nil, "", http.StatusInternalServerError).Wrap(nErr)
+	}
+
+	oldAck.AcknowledgedAt = 0
+
+	a.Srv().Go(func() {
+		a.sendAcknowledgementEvent(model.WebsocketEventAcknowledgementRemoved, oldAck, post)
+	})
+
+	return nil
+}
+
+func (a *App) GetAcknowledgementsForPost(postID string) ([]*model.PostAcknowledgement, *model.AppError) {
+	_, err := a.GetSinglePost(postID, false)
+	if err != nil {
+		return nil, err
+	}
+
+	acknowledgements, nErr := a.Srv().Store().PostAcknowledgement().GetForPost(postID)
+	if nErr != nil {
+		return nil, model.NewAppError("GetAcknowledgementsForPost", "app.acknowledgement.getforpost.get.app_error", nil, "", http.StatusInternalServerError).Wrap(nErr)
+	}
+
+	return acknowledgements, nil
+}
+
+func (a *App) GetAcknowledgementsForPostList(postList *model.PostList) (map[string][]*model.PostAcknowledgement, *model.AppError) {
+	acknowledgements, err := a.Srv().Store().PostAcknowledgement().GetForPosts(postList.Order)
+
+	if err != nil {
+		return nil, model.NewAppError("GetPostAcknowledgementsForPostList", "app.acknowledgement.get.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+	}
+
+	acknowledgementsMap := make(map[string][]*model.PostAcknowledgement)
```

Comment:
Can we pre-allocate this map based on the size of `acknowledgements`?

Reply:
`acknowledgmentsMap` is actually acknowledgments by post and not all posts have acknowledgements

e.g let's say we have just 10 acks all for one post but we ask acks for 200 posts, the map would have one key (the post id) with a slice of 10 acks in that key.

```go
map[string]*[]PostAcknowledgements{
   [post1.Id]: [10]PostAcknoledgements,
}
```

Reply:
Aha, I misread that! Thank you for clarifying :)